Visualization for Meshes / Surfaces / Solids in Preview Bubble and Watch Node

samuelcwl opened the pull request:
### Purpose

This PR is meant to address the _MAGN-10809_.

Previously, when users selected the nodes within a preview bubble or a Watch Node, all that they will see is the label on the HelixWatch3DPreview. The highlighting of the nodes in the HelixWatch3DPreview is only toggled on when the entire node is selected:

![10809old](https://cloud.githubusercontent.com/assets/16283396/19763668/b15d8d0e-9c71-11e6-9119-d2bfc6433a71.gif)

After this implementation, users will be able to select individual nodes in the TreeViewItem and see the highlighting/ visualization of the specific node which they had selected:

![10809new](https://cloud.githubusercontent.com/assets/16283396/19763670/b5be2250-9c71-11e6-94f3-cabe57e90122.gif)

One limitation, however, is that this preview ability is currently only constrained to Meshes, Surfaces and Solids. Lines, Curves and Points currently are not supported in this implementation as they are grouped as one object each (i.e. all the lines will be collapsed into a `Lines` object and so on) in the `Model3DDictionary` within HelixWatch3DViewModel. Due to the lack of granularity, individual lines when selected, cannot be teased out and highlighted.
### Declarations

Check these if you believe they are true
- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Follows [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) standards (_No change of public methods or types etc.._)
### Reviewers

@ke-yu 
### FYIs

@riteshchandawar

aparajit-pratap commented:
@samuelcwl thanks for implementing this! Looks good, please add a test case as discussed and then it's good to go.

mjkkirschner commented:
@samuelcwl some meshes are also stored in the same object as well, this may fail for those cases won't it?

samuelcwl commented:
@mjkkirschner, yes - this will probably fail for those cases as well. The issue lies in the way that those objects are initially built. Selection and de-selection, with this implementation, can only work for separate objects in HelixWatch3D.

samuelcwl commented:
Fixed some UX issues - including:

1. When the parent node is selected, and a node in the list is selected, the parent node is not de-selected.
2. When a node in the list is selected, and when user clicks elsewhere on canvas, the highlighting in the HelixWatch3D is turned off - even though the labels have not been toggled off or the node deselected.

+ Added Test Cases

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:
```diff
@@ -1366,6 +1385,35 @@ public override void ClearPathLabel(string path)
+
+            var nodePath = path.Contains(':') ? path.Remove(path.IndexOf(':')) : path;
+
+            if (DynamoSelection.Instance.Selection.Any())
```
is this check necessary?

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs:
```diff
@@ -609,9 +609,28 @@ protected override void SelectionChangedHandler(object sender, NotifyCollectionC
+
+                    var modelValues = geometryModels.Select(x => x.Value);
+
+                    foreach (GeometryModel3D g in modelValues)
```
Isn't this a double enumeration over the list? Maybe we can simply enumerate once over `geometryModels` in the `foreach` loop?

samuelcwl commented:
@aparajit-pratap , made recommended fixes.

aparajit-pratap commented:
Thanks @samuelcwl LGTM!
